pkg/diff/policy_sql_generator.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	return &policySQLVertexGenerator{
+		table:                  table,
+		newSchemaColumnsByName: buildSchemaObjByNameMap(table.Columns),
+		oldTable:               oldTable,
+		oldSchemaColumnsByName: oldSchemaColumnsByName,
+	}, nil
+}
+
+func (psg *policySQLVertexGenerator) Add(p schema.Policy) ([]Statement, error) {
+	sb := strings.Builder{}
+	sb.WriteString(fmt.Sprintf("CREATE POLICY %s ON %s", p.EscapedName, psg.table.GetFQEscapedName()))
+
+	typeModifier := "RESTRICTIVE"
+	if p.IsPermissive {
+		typeModifier = "PERMISSIVE"
+	}
+	sb.WriteString(fmt.Sprintf("\n\tAS %s", typeModifier))
+
+	cmdSQL, err := policyCharToSQL(p.Cmd)
+	if err != nil {
+		return nil, err
+	}
+	sb.WriteString(fmt.Sprintf("\n\tFOR %s", cmdSQL))
+
+	sb.WriteString(fmt.Sprintf("\n\tTO %s", strings.Join(p.AppliesTo, ", ")))
+
+	if p.UsingExpression != "" {
+		sb.WriteString(fmt.Sprintf("\n\tUSING (%s)", p.UsingExpression))
+	}
+	if p.CheckExpression != "" {
+		sb.WriteString(fmt.Sprintf("\n\tWITH CHECK (%s)", p.CheckExpression))
+	}
+
+	hazard := migrationHazardRestrictivePolicyAdded
+	if p.IsPermissive {
+		hazard = migrationHazardPermissivePolicyAdded
+	}
+
+	return []Statement{{
+		DDL:         sb.String(),
+		Timeout:     statementTimeoutDefault,
+		LockTimeout: lockTimeoutDefault,
+		Hazards:     []MigrationHazard{hazard},
+	}}, nil
+}
+
+func policyCharToSQL(c schema.PolicyCmd) (string, error) {
+	switch c {
+	case schema.SelectPolicyCmd:
+		return "SELECT", nil
+	case schema.InsertPolicyCmd:
+		return "INSERT", nil
+	case schema.UpdatePolicyCmd:
+		return "UPDATE", nil
+	case schema.DeletePolicyCmd:
+		return "DELETE", nil
+	case schema.AllPolicyCmd:
+		return "ALL", nil
+	default:
+		return "", fmt.Errorf("unknown policy command: %v", c)
+	}
+}
+
+func (psg *policySQLVertexGenerator) Delete(p schema.Policy) ([]Statement, error) {
+	hazard := migrationHazardRestrictivePolicyRemoved
+	if p.IsPermissive {
+		hazard = migrationHazardPermissivePolicyRemoved
+	}
+	return []Statement{{
+		DDL:         fmt.Sprintf("DROP POLICY %s ON %s", p.EscapedName, psg.table.GetFQEscapedName()),
+		Timeout:     statementTimeoutDefault,
+		LockTimeout: lockTimeoutDefault,
+		Hazards:     []MigrationHazard{hazard},
+	}}, nil
+}
+
+func (psg *policySQLVertexGenerator) Alter(diff policyDiff) ([]Statement, error) {
+	oldCopy := diff.old
+
+	// alterPolicyParts represents the set of strings to include in the ALTER POLICY ... ON TABLE ... statement
+	var alterPolicyParts []string
+
+	if !cmp.Equal(oldCopy.AppliesTo, diff.new.AppliesTo) {
+		alterPolicyParts = append(alterPolicyParts, fmt.Sprintf("TO %s", strings.Join(diff.new.AppliesTo, ", ")))
+		oldCopy.AppliesTo = diff.new.AppliesTo
+	}
+
+	if oldCopy.UsingExpression != diff.new.UsingExpression && diff.new.UsingExpression != "" {
+		// Weirdly, you can't actually drop a "USING EXPRESSION" clause from an ALL policy even though you
+		// can have an ALL policy with only a check expression.
```

Comment:
shouldn't there be an ErrNotImplemented somewhere to cover this case?

Reply:
Ah i see that's handled by if diff.old isn't mutated to match diff.new

Reply:
Yep. That's the "canonical" pattern I'm going for. It makes "ErrNotImplemented" free for any feature we inherently haven't explicitly coded support for.